[docs] Document CockroachDB sink target behavior in the JDBC connector page (#7636)
debezium/debezium: documentation/modules/ROOT/pages/connectors/jdbc.adoc

* xref:jdbc-idempotent-writes[Idempotent writes]
* xref:jdbc-schema-evolution[Schema evolution]
* xref:jdbc-singlestore-targets[SingleStore sink targets]

> * xref:jdbc-cockroachdb-targets[CockroachDB sink targets]

Reviewer: ```suggestion
ifdef::community[]
* xref:jdbc-cockroachdb-targets[CockroachDB sink targets]
endif::community[]
```
I'm assuming that in the near term the CockroachDB target is not available in the product edition, yes?

Author: Applied in 07e0bbd

SingleStore supports a subset of WKT geospatial values.
The connector writes `POINT`, `LINESTRING`, and `POLYGON` WKB values to SingleStore as WKT strings.

// Type: concept
// Title: Using CockroachDB as a sink target

> // ModuleID: debezium-jdbc-connector-using-cockroachdb-as-a-sink-target

Reviewer: ```suggestion
ifdef::community[]
```
If CockroachDB is not available in the product edition, these annotation comments should be removed and replaced with the `community` conditional delimiters here and at the end of the section, before the current L97.

[[jdbc-cockroachdb-targets]]
=== CockroachDB sink targets

> To write to CockroachDB, configure xref:jdbc-property-connection-url[`connection.url`] with a PostgreSQL JDBC URL, for example, `jdbc:postgresql://<_host_>:26257/<_database_>`.

Reviewer: ```suggestion
To enable the JDBC connector to write to CockroachDB, you must specify a PostgreSQL JDBC URL in the connector's xref:jdbc-property-connection-url[`connection.url`] property, for example, `jdbc:postgresql://<_host_>:26257/<_database_>`.
```

CockroachDB is wire compatible with PostgreSQL and uses the PostgreSQL JDBC driver that is included in the {prodname} JDBC connector archive.

> CockroachDB targets support idempotent writes through the same xref:jdbc-property-insert-mode[`insert.mode=upsert`] setting that the connector uses for other databases.

Reviewer: ```suggestion
As a sink target, CockroachDB supports idempotent writes through the same xref:jdbc-property-insert-mode[`insert.mode=upsert`] setting that the connector uses for other databases.
```

For CockroachDB, the connector generates `INSERT ... ON CONFLICT ... DO UPDATE SET ...` statements.

> CockroachDB runs transactions at `SERIALIZABLE` isolation and reports transient transaction conflicts with SQLSTATE 40001.

Reviewer: ```suggestion
CockroachDB uses the `SERIALIZABLE` transaction isolation level. 
As a result, if CockroachDB detects a conflict between concurrent transactions, it might temporarily abort one of the transactions and return SQLSTATE `40001`.
```

The connector treats these conflicts as retriable and retries the flush according to the xref:jdbc-property-flush-max-retries[`flush.max.retries`] and xref:jdbc-property-flush-retry-delay-ms[`flush.retry.delay.ms`] settings.
Setting `flush.max.retries` to `0` disables the retries so that a conflicting flush fails immediately.

For CockroachDB targets, {prodname} uses PostgreSQL-compatible mappings for most data types, with the following differences:

* `io.debezium.data.Json` logical values and Kafka Connect `MAP` schema values are written to the CockroachDB `jsonb` column type.
* `io.debezium.data.Xml` logical values and propagated range, `macaddr`, and `cidr` column types are written as `text`, because CockroachDB does not provide those types.

> * Propagated `money` column types are written as `decimal(19,2)`.

Reviewer: ```suggestion
* Propagated `money` column types are written as `decimal(19,2)` types.
```

* `io.debezium.data.geometry.Geometry`, `io.debezium.data.geometry.Geography`, and `io.debezium.data.geometry.Point` logical values are written to the built-in CockroachDB `geometry` and `geography` types; no PostGIS extension schema is required.

> * `io.debezium.data.FloatVector` logical values are written to the CockroachDB `vector` column type, because CockroachDB has no `halfvec` type. `io.debezium.data.SparseDoubleVector` is not supported.

Reviewer: ```suggestion
* `io.debezium.data.FloatVector` logical values are written to the CockroachDB `vector` column type, because CockroachDB has no `halfvec` type.
* 
`io.debezium.data.SparseDoubleVector` is not supported.
```

> 

Reviewer: ```suggestion
endif::community[]
```
See the note on L75.

Author: Applied in 07e0bbd

{prodname} `io.debezium.data.Json` logical values and Kafka Connect `MAP` schema values map to the StarRocks `JSON` column type, `ENUM` and `SET` values map to `STRING` columns, and `YEAR` values map to `INT` columns.
The StarRocks dialect does not support geospatial or vector logical types, and Kafka Connect `ARRAY` and `STRUCT` schema types are not mapped to the StarRocks `ARRAY` and `STRUCT` column types.

ifdef::community[]
[[jdbc-cockroachdb-targets]]
=== CockroachDB sink targets

To enable the JDBC connector to write to CockroachDB, you must specify a PostgreSQL JDBC URL in the connector's xref:jdbc-property-connection-url[`connection.url`] property, for example, `jdbc:postgresql://<_host_>:26257/<_database_>`.

> CockroachDB is wire compatible with PostgreSQL and uses the PostgreSQL JDBC driver that is included in the {prodname} JDBC connector archive.

Reviewer: ```suggestion
Because CockroachDB supports the PostgreSQL communication protocol, it works with the PostgreSQL JDBC driver that is included in the {prodname} JDBC connector archive.
```
I glossed over this on my first pass, but you should avoid the jargon term _wire compatible_ in favor of describing the behavior.

As a sink target, CockroachDB supports idempotent writes through the same xref:jdbc-property-insert-mode[`insert.mode=upsert`] setting that the connector uses for other databases.
For CockroachDB, the connector generates `INSERT ... ON CONFLICT ... DO UPDATE SET ...` statements.

CockroachDB uses the `SERIALIZABLE` transaction isolation level.
As a result, if CockroachDB detects a conflict between concurrent transactions, it might temporarily abort one of the transactions and return SQLSTATE `40001`.
The connector treats these conflicts as retriable and retries the flush according to the xref:jdbc-property-flush-max-retries[`flush.max.retries`] and xref:jdbc-property-flush-retry-delay-ms[`flush.retry.delay.ms`] settings.
Setting `flush.max.retries` to `0` disables the retries so that a conflicting flush fails immediately.

> For CockroachDB targets, {prodname} uses PostgreSQL-compatible mappings for most data types, with the following differences:

Reviewer: ```suggestion
For CockroachDB targets, {prodname} uses PostgreSQL-compatible mappings for most data types, with the following exceptions:
```

SingleStore maps {prodname} `io.debezium.data.Json` logical values to `json`.
Kafka Connect `MAP` schema values are also written to SingleStore `json` columns.
{prodname} `io.debezium.data.geometry.Geometry` and `io.debezium.data.geometry.Geography` logical values are written to SingleStore `geography` columns, and `io.debezium.data.geometry.Point` logical values are written to SingleStore `geographypoint` columns.

> For CockroachDB targets, {prodname} uses the PostgreSQL-compatible mappings for most dialect-specific logical types.

Reviewer: ```suggestion
When you configure the JDBC connector to use CockroachDB as a sink target, {prodname} uses the PostgreSQL-compatible mappings for most dialect-specific logical types.
For example, the connector uses the following mappings:

```

`io.debezium.data.Json` logical values and Kafka Connect `MAP` schema values are written to CockroachDB `jsonb` columns.
`io.debezium.data.Xml` logical values are written to `text` columns because CockroachDB has no XML type.
Geometry, geography, and point values are written to the built-in CockroachDB `geometry` and `geography` types without a PostGIS extension schema.

> `io.debezium.data.FloatVector` logical values are written to `vector` columns; sparse double vectors are not supported.

Reviewer: ```suggestion
* `io.debezium.data.Json` logical values and Kafka Connect `MAP` schema values are written to CockroachDB `jsonb` columns.
* `io.debezium.data.Xml` logical values are written to `text` columns because CockroachDB has no XML type.
* Geometry, geography, and point values are written to the built-in CockroachDB `geometry` and `geography` types without a PostGIS extension schema.
* `io.debezium.data.FloatVector` logical values are written to `vector` columns; sparse double vectors are not supported.
```
